control/discovery.py

```diff
@@ -327,17 +327,36 @@ def __init__(self, config):
             assert 0
         self.logger.info(f"discovery addr: {self.discovery_addr} port: {self.discovery_port}")
 
+        self.sock = None
+        self.passed_enter = False
         self.conn_vals = {}
         self.connection_counter = 1
         self.selector = selectors.DefaultSelector()
 
     def __enter__(self):
+        self.passed_enter = True
         return self
 
     def __exit__(self, exc_type, exc_value, traceback):
         if self.omap_state:
             self.omap_state.cleanup_omap()
             self.omap_state = None
+        try:
+            for key in self.conn_vals:
+                self.conn_vals[key].connection.close()
+        except Except as ex:
+            pass
+        self.conn_vals = {}
+        if self.selector:
+            self.selector.close()
+            self.selector = None
+        if self.sock:
+            try:
+                self.sock.close()
+            except Exception as ex:
+                pass
+            self.sock = None
+        self.passed_enter = False
```

Comment:
how about closing down the networking in steps, making sure the connection closed is no longer registered to selector?


```suggestion
        # Check the networking is initialized
         if not self.selector:
             return
        # Step 1, close all client connections
        try:
            for key in self.conn_vals:
                self.selector.unregister(self.conn_vals[key].connection)
                self.conn_vals[key].connection.close()
        except Except as ex:
            pass
        self.conn_vals = {}
        
        # Step 2 close the server socket
        if self.sock:
            try:
                self.selector.unregister(self.sock)
                self.sock.close()
            except Exception as ex:
                pass
            self.sock = None
            
        # Step 3 close the selector
        self.selector.close()
        self.selector = None
```

Reply:
I didn't touch the selector code. If you think there is a problem there you can open an issue.